crates/util/python-utils/src/taproot.rs

```diff
@@ -0,0 +1,207 @@
+    Emitter,
+};
+use bdk_wallet::{
+    bitcoin::{key::Parity, Address, AddressType, XOnlyPublicKey},
+    KeychainKind, Wallet,
+};
+use musig2::KeyAggContext;
+use pyo3::prelude::*;
+
+use crate::{
+    constants::{CHANGE_DESCRIPTOR, DESCRIPTOR, NETWORK},
+    error::Error,
+    parse::parse_xonly_pk,
+};
+
+/// Extracts the public key from a Taproot address.
+pub(crate) trait ExtractP2trPubkey {
+    /// Extracts the public key from a Taproot address.
+    fn extract_p2tr_pubkey(&self) -> Result<XOnlyPublicKey, Error>;
+}
+
+impl ExtractP2trPubkey for Address {
+    fn extract_p2tr_pubkey(&self) -> Result<XOnlyPublicKey, Error> {
+        match self.address_type() {
+            Some(AddressType::P2tr) => {}
+            _ => return Err(Error::NotTaprootAddress),
+        }
+
+        let script_pubkey = self.script_pubkey();
+
+        Ok(XOnlyPublicKey::from_slice(&script_pubkey.as_bytes()[2..]).expect("valid pub key"))
+    }
+}
+
+/// A simple Taproot-enable wallet.
+///
+/// # Note
+///
+/// This uses the
+pub(crate) fn taproot_wallet() -> Result<Wallet, Error> {
+    Ok(Wallet::create(*DESCRIPTOR, *CHANGE_DESCRIPTOR)
+        .network(NETWORK)
+        .create_wallet_no_persist()
+        .map_err(|_| Error::Wallet))?
+}
+
+/// Syncs a wallet with the network using `bitcoind` as the backend.
+///
+/// # Note
+///
+/// This function should be only used with Regtest.
+pub(crate) fn sync_wallet(wallet: &mut Wallet, rpc_client: Client) -> Result<(), Error> {
+    let wallet_tip = wallet.latest_checkpoint();
+    let mut emitter = Emitter::new(&rpc_client, wallet_tip, 0);
+    while let Some(block) = emitter.next_block().expect("valid block") {
+        let height = block.block_height();
+        let connected_to = block.connected_to();
+        wallet
+            .apply_block_connected_to(&block.block, height, connected_to)
+            .map_err(|_| Error::BitcoinD)?
+    }
+    Ok(())
+}
+
+/// Creates a new `bitoind` RPC client.
+pub(crate) fn client(
+    url: &str,
+    rpc_cookie: Option<&PathBuf>,
+    rpc_user: Option<&str>,
+    rpc_pass: Option<&str>,
+) -> Result<Client, Error> {
+    Ok(Client::new(
+        url,
+        match (rpc_cookie, rpc_user, rpc_pass) {
+            (None, None, None) => Auth::None,
+            (Some(path), _, _) => Auth::CookieFile(path.clone()),
+            (_, Some(user), Some(pass)) => Auth::UserPass(user.into(), pass.into()),
+            (_, Some(_), None) => panic!("rpc auth: missing rpc_pass"),
+            (_, None, Some(_)) => panic!("rpc auth: missing rpc_user"),
+        },
+    )
+    .map_err(|_| Error::RpcClient))?
+}
+
+/// MuSig2 aggregates public keys into a single public key.
+///
+/// # Note
+///
+/// These should all be X-only public keys, assuming that all are [`Parity::Even`].
+pub(crate) fn musig_aggregate_pks_inner(pks: Vec<XOnlyPublicKey>) -> Result<XOnlyPublicKey, Error> {
+    let pks: Vec<(XOnlyPublicKey, Parity)> = pks.into_iter().map(|pk| (pk, Parity::Even)).collect();
```

Comment:
No, we can't because `KeyAggContext` takes an iterator of anything that can be `Into<Point>` and the `From`/`Into` is not for `XOnlyPublicKey`s but actually for `(XOnlyPublicKey, Parity)`s.

See: https://docs.rs/secp/0.3.0/secp/struct.Point.html#impl-From%3CPoint%3E-for-(XOnlyPublicKey,+Parity)

Which makes total sense.

Reply:
Oh I see. There is another method on `XOnlyPublicKey` called `.public_key(parity)`, I think which you can use instead of the tuple. But this is fine too.

Reply:
Somehow I am getting a conflict with `PublicKey` types now while trying your suggestion.

Reply:
We can leave this as is and merge this. I'll see if I can get it to work when I start working on the fn tests.